### ISSUE TITLE
Option to prevent default in domSource.select().events()

### ISSUE THREAD

**Issue by grozen** [cyclejs/cyclejs#65]
Cycle allows turning virtually any DOM event into a stream to be consumed by intents. However, for some events this proves pointless as the event's default handling is disruptive.

For example, suppose I have a form and would like to receive a stream of submissions on it, I could do something such as:

``` javascript
Cycle.h('form#search', {onsubmit: 'searchSubmits$'}, [
  Cycle.h('input#q', {autocomplete: 'off', name: 'query', type: 'text', placeholder: 'Add term'})
])
```

While this would indeed create a stream, the form would still actually navigate to its target upon submission, thus denying me the ability to make sensible use of the stream.

What I would like to have (and will gladly make a reality) is something like this:

``` javascript
Cycle.h('form#search', {onsubmit: 'searchSubmits$!'}, [
  Cycle.h('input#q', {autocomplete: 'off', name: 'query', type: 'text', placeholder: 'Add term'})
])
```

Which will basically retain the same behavior, except it will also call `preventDefault` for the submit event, allowing use of the stream.

I am, of course, entirely open to suggestions on how to actually signify this behavior. Being a Ruby buff I am all for adding bangs to the end of things to signify anything drastic, but this is not necessarily the best way.

**Comment by staltz** [comment]
Yes this should be fixed. The default Cycle behavior should be to preventDefault when using onsubmit: 'searchSubmits$' (without bang).
 
I am curious if there is any case at all where we want to allow the event's default instead of preventing it.
 
Hence, I'll fix this soon.

**Comment by grozen** [comment]
Actually, the very first example [here](https://developer.mozilla.org/en/docs/Web/API/event.preventDefault) seems to fit the bill. People may still want their checkboxes to toggle even if the clicks are turned into a stream, no?

**Comment by staltz** [comment]
Good point. Can you experiment whether the following approach solves your problem?

```javascript
var View = Cycle.createView(function () {
  return {vtree$: h('form#search', {onsubmit: 'searchSubmits$'})};
});

var Intent = Cycle.createIntent(function (view) {
  return {submitForm$: view.get('searchSubmits$')
    .map(function (ev) { ev.preventDefault(); return ev; })
  };
});
```

In order words, manually putting `preventDefault` in the Intent.

**Comment by staltz** [comment]
@grozen I'm getting confident that you can just put the `ev.preventDefault()` in your Intents. Here is an example I put together: http://jsfiddle.net/staltz/wdxs67x9/2/

**Comment by grozen** [comment]
Yup, that hits the mark. Thanks!

Still, this feels a little "wrong", if you ask me. Maybe what's bothering me is that it's odd for the intent to do anything with the events other than extract data from them

**Comment by staltz** [comment]
Yes you're right, it feels wrong because this is a side effect: we are mutating the event instead of mapping it.

In the end of the day, the DOM and its events was made as a mutable data structure, and as much as we want to hide it, it will stay mutable and "dirty". I can see these alternatives:

- Put `ev.preventDefault()` in Intents, as we did above
- Build a framework feature to apply `preventDefault()` when the event carrier is suffixed with bang (or some other convention)
- Call the `preventDefault()` in a `.doOnNext()` operator in the Intent (essentially the same as first alternative)
- Make a custom element that wraps the native checkbox, with `preventDefault()` always applied

I prefer either alternatives 3 or 4. In 3, we use a side effect, but at least we make it explicit, because that is what doOnNext() is made for. Another way of hiding side effects is alternative 4, which I think is quite elegant.
@grozen

**Comment by grozen** [comment]
Sounds sound. Thanks for taking the time to help me sort this out!

**Comment by btipling** [comment]
Using `doOnNext` seems reasonable if you're using rxjs but if you're using xstream the analogue would a `debug` but in this case it's not a debugging activity so that doesn't seem right either.

**Comment by Widdershin** [comment]
Why don't we make it an optional argument to events? `DOM.select('.add').events('click', {preventDefault: true})`

**Comment by btipling** [comment]
So @staltz explained to me on gitter that [snapdom has hooks](https://github.com/snabbdom/snabbdom#hooks) you can use for this. There's an example on the readme:

```javascript
h('div.row', {
  key: movie.rank,
  hook: {
    insert: (vnode) => { movie.elmHeight = vnode.elm.offsetHeight; }
  }
});
```

Seems like that's the way to go for something like `preventDefault`. I have already written a driver but I'll probably move over to this.

**Comment by abaco** [comment]
@Widdershin Your proposal looks convenient and readable, but is it in accordance to cycle's principles? Wouldn't it violate the rule that side-effects are only allowed in drivers? I'm really asking this, it's not a rethorical question.

So my question is: is your `{preventDefault: true}` a side-effect? I think one may argue that it is or that it isn't, with equal merit. I'd say it's *not* a side effect. It's true that `ev.preventDefault()` will be called at some point, and that's technically a side effect, but so is `el.addEventListener('click', ...)`, and `.events('click')` may or may not call it, but that's just an implementation detail. The semantics of `.events('click')` is not to cause a side effect, and I believe that's also the case of `{preventDefault: true}`.  What if a new browser API was added, something like `el.addEventListener('click', {preventDefault: true})`? In that case cycle wouldn't need to mutate the event, nor to cause any side-effect. But, would that implementation detail change the semantics of `.events('click', {preventDefault: true})`. I believe it wouldn't.

In short, I believe your proposal should be accepted if there's a general agreement that it doesn't involve a side-effect, from the point of view of the semantics of the API (not of the low-level implementation).

**Comment by staltz** [comment]
I actually like the `.events('click', {preventDefault: true})` idea. The fact that a preventDefault driver works now is barely possible. It works just because the scheduler in xstream is recursive/synchronous only, so the preventDefault is called synchronously. However, if we would change the scheduling in xstream, or if the user is using Cycle with another stream library, this driver solution would be fragile in regard to the scheduling used.

When it comes to "is it an effect", I think we can give it semantics like "I want to listen for *prevented* click events, so I call `.events('click', {preventDefault: true})`. Of course we have to consider effects from the declarative perspective since the virtual DOM itself is a declarative abstraction that hides away operations like `element.appendChild()`.

So 👍 for this idea from my side.

**Comment by abaco** [comment]
What if one wants to `preventDefault` conditionally? Then there's no other option than using a driver - or is there?

Besides `preventDefault` there are other mutations of events that don't seem to fit in this nice  API, like those of `ev.dataTransfer`. What shall we do with them? Will there ever be a safe way to perform such side-effects?

**Comment by Widdershin** [comment]
@abaco if the condition is solely based on the event, you could pass a function instead. `.events('keydown', {preventDefault: ev => ev.code === 'space'})`

If the conditional is based on other parts of your application state, you would need to use the driver approach.

I think we should consider `dataTransfer` as a separate issue, as it seems like a different and more complex abstraction would be needed.

**Comment by abaco** [comment]
@Widdershin, don't get me wrong, I think your proposal solves 90% of the cases that currently require event mutation, so let's do it. For the remaining 10% I'd be happy to use a driver.

But @staltz has raised a new issue, namely that drivers that mutate DOM events are basically working by chance. That means we should *never* use the driver approach, doesn't it? So how are we going to deal with complex conditional event mutations?

**Comment by staltz** [comment]
This issue is super old, but quite easy to do. If anyone wants to do a first PR for Cycle.js, this would be a great starting point, we can guide you in the PR review.

**Comment by grozen** [comment]
Hmm, seems I'm late to the party.
If you're still up for volunteers I'll gladly give this a jab, though to be completely honest I couldn't TypeScript myself out of a wet paper bag.

**Comment by Widdershin** [comment]
Sure thing @grozen, have a go. I will review your pull request and help you get it merged. 🙂 👍

**Comment by grozen** [comment]
This is still coming along if anyone was concerned. I feed it what scraps of time I can and will hopefully be done (or rather, done enough to submit a pull request) soon.

**Comment by jvanbruegge** [comment]
Lol, I did not know about this issue, but came to the same conclusion: https://github.com/cyclejs/cyclejs/pull/603
Should I add the ability to pass a function?

**Comment by grozen** [comment]
@jvanbruegge since that is part of what's been discussed here, I'm allowing either a boolean or a function so that should be taken care of. I'll do my best to wrap this up in the coming days.

**Comment by grozen** [comment]
Also @Widdershin, I just noticed that I actually forgot that I was supposed to "make commit" instead of just committing while I work. How do I go about rectifying that? Everything is currently in a fork so I could essentially "start fresh" and copy everything over.

**Comment by staltz** [comment]
I'll merge #603 because it's useful by itself as a small "atomic" PR. We can improve the preventDefault features later on too. @grozen try a `git rebase -i`. The make commit is not important **if** you commit following the convention correctly.

**Comment by staltz** [comment]
This is done in cycle/dom 17.2 thanks to @jvanbruegge :)

**Comment by grozen** [comment]
@jvanbruegge the reason the tests keep looping over is that the when the default event handling for form submission is not prevented the page that runs the tests is reloaded and everything starts over.
I know because I tackled this as part of #610, which we can mix and match with this if it's deemed important enough.
